controller/worker/domain_migration/domain_migration.go

```diff
@@ -0,0 +1,254 @@
+}
+
+func (c *context) HandleDomainMigration(job *que.Job) (err error) {
+	log := c.logger.New("fn", "HandleDomainMigration")
+	log.Info("handling domain migration", "job_id", job.ID, "error_count", job.ErrorCount)
+
+	var dm *ct.DomainMigration
+	if err := json.Unmarshal(job.Args, &dm); err != nil {
+		log.Error("error unmarshaling job", "err", err)
+		return err
+	}
+
+	log = log.New("domain_migration", dm.ID)
+
+	if err := c.db.QueryRow("SELECT old_domain, domain, old_tls_cert, tls_cert, created_at, finished_at FROM domain_migrations WHERE migration_id = $1", dm.ID).Scan(&dm.OldDomain, &dm.Domain, &dm.OldTLSCert, &dm.TLSCert, &dm.CreatedAt, &dm.FinishedAt); err != nil {
+		log.Error("error fetching postgres record", "err", err)
+		c.createEvent(dm, err)
+		return err
+	}
+
+	if dm.FinishedAt != nil {
+		// Already done
+		return nil
+	}
+
+	// Generate TLS Cert if not already present
+	var cert *tlscert.Cert
+	if dm.TLSCert == nil {
+		var err error
+		cert, err = c.generateTLSCert(dm.ID, dm.Domain)
+		if err != nil {
+			log.Error("error generating TLS cert", "err", err)
+			c.createEvent(dm, err)
+			return err
+		}
+	} else {
+		cert = dm.TLSCert
+	}
+	oldCert := dm.OldTLSCert
+
+	if err := c.maybeDeployController(oldCert, cert, dm.OldDomain, dm.Domain); err != nil {
+		log.Error("error deploying controller", "err", err)
+		c.createEvent(dm, err)
+		return err
+	}
+
+	if err := c.maybeDeployRouter(oldCert, cert, dm.OldDomain, dm.Domain); err != nil {
+		log.Error("error deploying router", "err", err)
+		c.createEvent(dm, err)
+		return err
+	}
+
+	if err := c.maybeDeployDashboard(oldCert, cert, dm.OldDomain, dm.Domain); err != nil {
+		log.Error("error deploying dashboard", "err", err)
+		c.createEvent(dm, err)
+		return err
+	}
+
+	if err := c.createMissingRoutes(oldCert, cert, dm.OldDomain, dm.Domain); err != nil {
+		log.Error("error creating missing routes", "err", err)
+		c.createEvent(dm, err)
+		return err
+	}
+
+	if err := c.db.QueryRow("UPDATE domain_migrations SET finished_at = now() RETURNING finished_at").Scan(&dm.FinishedAt); err != nil {
+		log.Error("error setting finished_at", "err", err)
+		c.createEvent(dm, err)
+		return err
+	}
+
+	log.Info("domain migration complete")
+
+	c.createEvent(dm, nil)
+
+	return nil
+}
+
+func (c *context) generateTLSCert(id, domain string) (*tlscert.Cert, error) {
+	hosts := []string{
+		domain,
+		fmt.Sprintf("*.%s", domain),
+	}
+	cert, err := tlscert.Generate(hosts)
+	if err != nil {
+		return nil, err
+	}
+	if _, err := c.db.Query("UPDATE domain_migrations SET tls_cert = $1 WHERE migration_id = $2", cert, id); err != nil {
+		return nil, err
+	}
+	return cert, nil
+}
+
+func dupRelease(release *ct.Release) *ct.Release {
+	return &ct.Release{
+		ArtifactID: release.ArtifactID,
+		Env:        release.Env,
+		Meta:       release.Meta,
+		Processes:  release.Processes,
+	}
+}
+
+func (c *context) maybeDeployController(oldCert, cert *tlscert.Cert, oldDomain, newDomain string) error {
```

Comment:
This should check for a deployment in progress and wait for it to complete first